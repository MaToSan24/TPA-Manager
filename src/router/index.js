import { createRouter, createWebHistory } from "vue-router"
import Home from "../views/Home.vue"

const routes = [
    {
        path: "/",
        name: "home",
        component: Home,
    },
    {
        path: "/create",
        name: "create",
        component: () => import("../views/Create.vue"),
    },
    {
        path: "/visualization/:courseId?/:projectId?",
        name: "visualization",
        component: () => import("../views/Visualization.vue"),
    },
    {
        path: "/edition/:courseId?/:projectId?",
        name: "edition",
        component: () => import("../views/Edition.vue"),
        props: true,
    },
    {
        path: "/catalogue",
        name: "catalogue",
        component: () => import("../views/Catalogue.vue"),
    },
    {
        path: "/:catchAll(.*)",
        name: "notFound",
        component: () => import("../views/NotFound.vue"),
    },
]

const router = createRouter({
    history: createWebHistory(), // Since this is a SPA, we need to manage this in the server by redirecting all the routes to index.html.
    routes,
})

export default router;