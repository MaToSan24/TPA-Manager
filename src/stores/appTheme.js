import { ref, computed } from 'vue'
import { defineStore } from 'pinia'
import { usePrimeVue } from 'primevue/config';

export const useAppThemeStore = defineStore('appTheme', () => {
  const PrimeVue = usePrimeVue();
  const appTheme = ref(localStorage.getItem("appTheme") || 'Light');

  const isDarkModeOn = computed(() => appTheme.value === 'Dark');
  
  function toggleTheme() {
    if (appTheme.value === 'Dark') {
      appTheme.value = 'Light';
      localStorage.setItem("appTheme", "Light");
      PrimeVue.changeTheme('arya-blue', 'lara-light-blue', 'theme-link', () => {});
    } else {
      appTheme.value = 'Dark';
      localStorage.setItem("appTheme", "Dark");
      PrimeVue.changeTheme('lara-light-blue', 'arya-blue', 'theme-link', () => {});
    }
  }
  
  return { appTheme, isDarkModeOn, toggleTheme }
});