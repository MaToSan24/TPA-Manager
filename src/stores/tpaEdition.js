import { defineStore } from 'pinia'
import { ref } from 'vue'
import _ from 'lodash'

export const useTpaEditionStore = defineStore('tpaEdition', () => {
  const originalTpa = ref(null)
  const modifiedTpa = ref(null)

  function setInitialTpaData(tpa) {
    originalTpa.value = _.cloneDeep(tpa)
    modifiedTpa.value = _.cloneDeep(tpa)
  }

  function saveTpaChanges() {
    originalTpa.value = modifiedTpa
  }

  function discardTpaChanges() {
    modifiedTpa.value = originalTpa
  }

  function getTpaField(fieldPath) {
    return _.get(modifiedTpa.value, fieldPath)
  }
  
  function updateTpaField(fieldPath, value) {
    _.set(modifiedTpa.value, fieldPath, value)
  }

  return { originalTpa, modifiedTpa, setInitialTpaData, saveTpaChanges, discardTpaChanges, getTpaField, updateTpaField }
})