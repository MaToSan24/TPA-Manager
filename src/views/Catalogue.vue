<script setup>
import { useRouter } from 'vue-router';

import Button from 'primevue/button';

const router = useRouter();
</script>

<template>
  <div class="card" style="display: grid; justify-items: center; justify-content: center; grid-template-rows: auto auto; position: relative; width: fit-content; left: 50%; transform: translateX(-50%);">
    <h1>TPs Catalogue View</h1>
    <div style="display: grid; grid-auto-flow: column; gap: 1rem;">
        <Button label="🏠 Home" @click="router.push({ name: 'home' })" />
        <Button label="🔙 Back" @click="router.go(-1)" />
    </div>
  </div>
</template>