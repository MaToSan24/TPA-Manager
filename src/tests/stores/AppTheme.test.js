import router from '@/router/index';
import PrimeVue from 'primevue/config';
import Catalogue from '@/views/Catalogue.vue';
import { mount } from '@vue/test-utils';
import { setActivePinia, createPinia } from 'pinia'
import { useAppThemeStore } from "@/stores/appTheme";

const pinia = createPinia();
let store;

describe("AppTheme store", () => {
    beforeEach(() => {
        setActivePinia(pinia)
        store = useAppThemeStore();

        // Manually add the <link> element to the <head> in the DOM for the tests to work
        const link = document.createElement('link');
        link.id = 'theme-link';
        link.href = '@/assets/themes/lara-light-blue/theme.css';
        document.head.appendChild(link);
    })
    
    const wrapper = mount(Catalogue, {
        global: {
            plugins: [router, pinia, PrimeVue],
        },
    });
    
    it('mounts correctly', () => {
        expect(wrapper.isVisible()).toBe(true);
    });
    
    test("should have the default theme", () => {
        expect(store.appTheme).toEqual("Light");
    });

    test("should toggle the theme", () => {
        store.toggleTheme();
        expect(store.appTheme).toEqual("Dark");
        expect(store.isDarkModeOn).toEqual(true);
        store.toggleTheme();
        expect(store.appTheme).toEqual("Light");
    });

    it("matches snapshot in light mode", () => {
        expect(wrapper.html()).toMatchSnapshot();
    });

    it("matches snapshot in dark mode", async () => {
        wrapper.find("button.right-0").trigger("click");
        await wrapper.vm.$nextTick(); // Wait for the component to update
        expect(wrapper.html()).toMatchSnapshot();
    });
});

