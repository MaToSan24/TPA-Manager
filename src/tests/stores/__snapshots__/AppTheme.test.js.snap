// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AppTheme store > matches snapshot in dark mode 1`] = `
"<button data-v-cbf9ab53=\\"\\" class=\\"p-button p-component p-button-icon-only p-button-secondary absolute ml-3 mt-2 left-0 top-0\\" type=\\"button\\" data-pc-name=\\"button\\" data-pc-section=\\"root\\" data-pc-severity=\\"secondary\\" data-pd-ripple=\\"true\\"><span class=\\"p-button-icon pi pi-home\\" data-pc-section=\\"icon\\"></span><span class=\\"p-button-label\\" data-pc-section=\\"label\\"> </span>
  <!---->
</button>
<button data-v-cbf9ab53=\\"\\" class=\\"p-button p-component p-button-icon-only p-button-secondary absolute ml-3 mt-2 left-0 top-0\\" type=\\"button\\" data-pc-name=\\"button\\" data-pc-section=\\"root\\" data-pc-severity=\\"secondary\\" style=\\"margin-left: 4.55rem !important;\\" data-pd-ripple=\\"true\\"><span class=\\"p-button-icon pi pi-arrow-left\\" data-pc-section=\\"icon\\"></span><span class=\\"p-button-label\\" data-pc-section=\\"label\\"> </span>
  <!---->
</button>
<button data-v-cbf9ab53=\\"\\" class=\\"p-button p-component p-button-icon-only p-button-secondary absolute mr-3 mt-2 right-0 top-0\\" type=\\"button\\" data-pc-name=\\"button\\" data-pc-section=\\"root\\" data-pc-severity=\\"secondary\\" data-pd-ripple=\\"true\\"><span data-pc-section=\\"icon\\" class=\\"p-button-icon pi pi-moon\\"></span><span class=\\"p-button-label\\" data-pc-section=\\"label\\"> </span>
  <!---->
</button>
<div data-v-cbf9ab53=\\"\\" class=\\"card\\">
  <h1 data-v-cbf9ab53=\\"\\" class=\\"text-center font-bold\\"><u data-v-cbf9ab53=\\"\\">📖 TPs Catalogue</u></h1>
  <div data-v-cbf9ab53=\\"\\" class=\\"flex flex-column m-5\\">
    <div data-v-cbf9ab53=\\"\\" class=\\"p-progress-spinner text-center\\" role=\\"progressbar\\" data-pc-name=\\"progressspinner\\" data-pc-section=\\"root\\"><svg class=\\"p-progress-spinner-svg\\" viewBox=\\"25 25 50 50\\" style=\\"animation-duration: 2s;\\" data-pc-section=\\"spinner\\">
        <circle class=\\"p-progress-spinner-circle\\" cx=\\"50\\" cy=\\"50\\" r=\\"20\\" fill=\\"none\\" stroke-width=\\"4\\" strokeMiterlimit=\\"10\\" data-pc-section=\\"circle\\"></circle>
      </svg></div>
    <h3 data-v-cbf9ab53=\\"\\" class=\\"text-center\\">Loading...</h3>
  </div>
  <div data-v-cbf9ab53=\\"\\" style=\\"display: grid; grid-auto-flow: column; gap: 1rem;\\"><button data-v-cbf9ab53=\\"\\" class=\\"p-button p-component\\" type=\\"button\\" aria-label=\\"🏠 Home\\" data-pc-name=\\"button\\" data-pc-section=\\"root\\" data-pd-ripple=\\"true\\">
      <!----><span class=\\"p-button-label\\" data-pc-section=\\"label\\">🏠 Home</span>
      <!---->
    </button><button data-v-cbf9ab53=\\"\\" class=\\"p-button p-component\\" type=\\"button\\" aria-label=\\"🔙 Back\\" data-pc-name=\\"button\\" data-pc-section=\\"root\\" data-pd-ripple=\\"true\\">
      <!----><span class=\\"p-button-label\\" data-pc-section=\\"label\\">🔙 Back</span>
      <!---->
    </button></div>
</div>"
`;

exports[`AppTheme store > matches snapshot in light mode 1`] = `
"<button data-v-cbf9ab53=\\"\\" class=\\"p-button p-component p-button-icon-only p-button-secondary absolute ml-3 mt-2 left-0 top-0\\" type=\\"button\\" data-pc-name=\\"button\\" data-pc-section=\\"root\\" data-pc-severity=\\"secondary\\" data-pd-ripple=\\"true\\"><span class=\\"p-button-icon pi pi-home\\" data-pc-section=\\"icon\\"></span><span class=\\"p-button-label\\" data-pc-section=\\"label\\"> </span>
  <!---->
</button>
<button data-v-cbf9ab53=\\"\\" class=\\"p-button p-component p-button-icon-only p-button-secondary absolute ml-3 mt-2 left-0 top-0\\" type=\\"button\\" data-pc-name=\\"button\\" data-pc-section=\\"root\\" data-pc-severity=\\"secondary\\" style=\\"margin-left: 4.55rem !important;\\" data-pd-ripple=\\"true\\"><span class=\\"p-button-icon pi pi-arrow-left\\" data-pc-section=\\"icon\\"></span><span class=\\"p-button-label\\" data-pc-section=\\"label\\"> </span>
  <!---->
</button>
<button data-v-cbf9ab53=\\"\\" class=\\"p-button p-component p-button-icon-only p-button-secondary absolute mr-3 mt-2 right-0 top-0\\" type=\\"button\\" data-pc-name=\\"button\\" data-pc-section=\\"root\\" data-pc-severity=\\"secondary\\" data-pd-ripple=\\"true\\"><span class=\\"p-button-icon pi pi-sun\\" data-pc-section=\\"icon\\"></span><span class=\\"p-button-label\\" data-pc-section=\\"label\\"> </span>
  <!---->
</button>
<div data-v-cbf9ab53=\\"\\" class=\\"card\\">
  <h1 data-v-cbf9ab53=\\"\\" class=\\"text-center font-bold\\"><u data-v-cbf9ab53=\\"\\">📖 TPs Catalogue</u></h1>
  <div data-v-cbf9ab53=\\"\\" class=\\"flex flex-column m-5\\">
    <div data-v-cbf9ab53=\\"\\" class=\\"p-progress-spinner text-center\\" role=\\"progressbar\\" data-pc-name=\\"progressspinner\\" data-pc-section=\\"root\\"><svg class=\\"p-progress-spinner-svg\\" viewBox=\\"25 25 50 50\\" style=\\"animation-duration: 2s;\\" data-pc-section=\\"spinner\\">
        <circle class=\\"p-progress-spinner-circle\\" cx=\\"50\\" cy=\\"50\\" r=\\"20\\" fill=\\"none\\" stroke-width=\\"4\\" strokeMiterlimit=\\"10\\" data-pc-section=\\"circle\\"></circle>
      </svg></div>
    <h3 data-v-cbf9ab53=\\"\\" class=\\"text-center\\">Loading...</h3>
  </div>
  <div data-v-cbf9ab53=\\"\\" style=\\"display: grid; grid-auto-flow: column; gap: 1rem;\\"><button data-v-cbf9ab53=\\"\\" class=\\"p-button p-component\\" type=\\"button\\" aria-label=\\"🏠 Home\\" data-pc-name=\\"button\\" data-pc-section=\\"root\\" data-pd-ripple=\\"true\\">
      <!----><span class=\\"p-button-label\\" data-pc-section=\\"label\\">🏠 Home</span>
      <!---->
    </button><button data-v-cbf9ab53=\\"\\" class=\\"p-button p-component\\" type=\\"button\\" aria-label=\\"🔙 Back\\" data-pc-name=\\"button\\" data-pc-section=\\"root\\" data-pd-ripple=\\"true\\">
      <!----><span class=\\"p-button-label\\" data-pc-section=\\"label\\">🔙 Back</span>
      <!---->
    </button></div>
</div>"
`;
